accumulator/batchproof.go

```diff
@@ -14,99 +14,126 @@ type BatchProof struct {
-// I don't think the binary.Write errors ever actually happen
-func (bp *BatchProof) ToBytes() []byte {
-	var buf bytes.Buffer
+/*
+Batchproof serialization is:
+4bytes numTargets
+4bytes numHashes
+[]Targets (8 bytes each)
+[]Hashes (32 bytes each)
+*/
 
+// Serialize a batchproof to a writer.
+func (bp *BatchProof) Serialize(w io.Writer) (err error) {
 	// first write the number of targets (4 byte uint32)
-	numTargets := uint32(len(bp.Targets))
-	if numTargets == 0 {
-		return nil
+	err = binary.Write(w, binary.BigEndian, uint32(len(bp.Targets)))
+	if err != nil {
+		return err
 	}
-	err := binary.Write(&buf, binary.BigEndian, numTargets)
+	// if there are no targets, finish early & don't write proofs
+	// if len(bp.Targets) == 0 {
+	// return
+	// }
+	// write out number of hashes in the proof
+	err = binary.Write(w, binary.BigEndian, uint32(len(bp.Proof)))
 	if err != nil {
-		fmt.Printf("huh %s\n", err.Error())
-		return nil
+		return
 	}
+
+	// write out each target
 	for _, t := range bp.Targets {
 		// there's no need for these to be 64 bit for the next few decades...
-		err := binary.Write(&buf, binary.BigEndian, t)
+		err = binary.Write(w, binary.BigEndian, t)
 		if err != nil {
-			fmt.Printf("huh %s\n", err.Error())
-			return nil
+			return
 		}
 	}
+
 	// then the rest is just hashes
 	for _, h := range bp.Proof {
-		_, err = buf.Write(h[:])
+		_, err = w.Write(h[:])
 		if err != nil {
-			fmt.Printf("huh %s\n", err.Error())
-			return nil
+			return
 		}
 	}
-
-	return buf.Bytes()
+	return
 }
 
-// ToString for debugging, shows the blockproof
-func (bp *BatchProof) SortTargets() {
-	sortUint64s(bp.Targets)
-}
+// TODO: could make this more efficient by not encoding as much empty stuff
 
-// ToString for debugging, shows the blockproof
-func (bp *BatchProof) ToString() string {
-	s := fmt.Sprintf("%d targets: ", len(bp.Targets))
-	for _, t := range bp.Targets {
-		s += fmt.Sprintf("%d\t", t)
-	}
-	s += fmt.Sprintf("\n%d proofs: ", len(bp.Proof))
-	for _, p := range bp.Proof {
-		s += fmt.Sprintf("%04x\t", p[:4])
-	}
-	s += "\n"
-	return s
+func (bp *BatchProof) SerializeSize() int {
+	// empty batchProofs are 4 bytes
+	// if len(bp.Targets) == 0 {
+	// 	return 4
+	// }
+	// 8B for numTargets and numHashes, 8B per target, 32B per hash
+	return 8 + (8 * (len(bp.Targets))) + (32 * (len(bp.Proof)))
 }
 
-// FromBytesBatchProof gives a block proof back from the serialized bytes
-func FromBytesBatchProof(b []byte) (BatchProof, error) {
-	var bp BatchProof
-
-	// if empty slice, return empty BatchProof with 0 targets
-	if len(b) == 0 {
-		return bp, nil
+// Deserialize gives a block proof back from the serialized bytes
+func (bp *BatchProof) Deserialize(r io.Reader) (err error) {
+	var numTargets, numHashes uint32
+	err = binary.Read(r, binary.BigEndian, &numTargets)
+	if err != nil {
+		if err == io.EOF && numTargets == 0 {
+			err = nil // EOF at the end is not an error...
```

Comment:
why is numTargets being compared to zero?  the [docs](https://golang.org/pkg/encoding/binary/#Read) say that it returns EOF only if nothing at all was read. so i think this extra condition should only be necessary to prevent partial reading?

Reply:
Yeah that shouldn't be there, taken out.  With reader.Read that can return EOFs with a successful read but not binary.Read.